network/netidentity.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// ProtocolConectionIdentityChallengeHeader is used to exchange IdentityChallenges
+const ProtocolConectionIdentityChallengeHeader = "X-Algorand-IdentityChallenge"
+
+// minimumProtocolVersion is used to evaluate if a peer's supported protocol
+// should include identityChallenge exchange
+var minimumProtocolVersion = [2]int64{2, 2}
+
+func shouldSupportIdentityChallenge(v string) bool {
+	maj, min, err := versionToMajorMinor(v)
+	if err != nil {
+		return false
+	}
+	if maj < minimumProtocolVersion[0] {
+		return false
+	}
+	if maj > minimumProtocolVersion[0] {
+		return true
+	}
+	if min < minimumProtocolVersion[1] {
+		return false
+	}
+	return true
+}
+
+type identityChallenge struct {
+	Key       crypto.PublicKey `codec:"pk"`
+	Challenge [32]byte         `codec:"c"`
+	Signature crypto.Signature `codec:"s"`
+}
+
+type identityChallengeResponse struct {
+	identityChallenge
+	ResponseChallenge [32]byte `codec:"rc"`
+}
+
+// NewIdentityChallenge creates an IdentityChallenge with randomized 32byte Challenge
+func NewIdentityChallenge(p crypto.PublicKey) identityChallenge {
+	c := identityChallenge{
+		Key:       p,
+		Challenge: [32]byte{},
+	}
+	crypto.RandBytes(c.Challenge[:])
+	return c
+}
+
+func (i identityChallenge) signableBytes() []byte {
+	return bytes.Join([][]byte{
```

Comment:
It's crypto best practices to prefix this with a couple bytes making the message even more clearly this message and not confusable for anything else. e.g. `[]byte("IC")` would be enough